packages/workflow/src/workflow.ts

```diff
@@ -1356,21 +1372,98 @@ export function upsertSearchAttributes(searchAttributes: SearchAttributes): void
+      upsertWorkflowSearchAttributes: {
+        searchAttributes: typedMapToPayloads<string, TypedSearchAttributeUpdateValue>(
+          typedSearchAttributePayloadConverter,
+          Object.fromEntries(searchAttributes.map(([k, v]) => [k.name, v]))
+        ),
+      },
+    });
 
-  activator.mutateWorkflowInfo((info: WorkflowInfo): WorkflowInfo => {
-    return {
-      ...info,
-      searchAttributes: {
-        ...info.searchAttributes,
-        ...searchAttributes,
+    activator.mutateWorkflowInfo((info: WorkflowInfo): WorkflowInfo => {
+      const newSearchAttributes: SearchAttributes = {};
+      for (const pair of searchAttributes) {
+        const [k, v] = pair;
+        if (v === null) {
+          // If the value is null (i.e. delete), set the value to an empty list.
+          newSearchAttributes[k.name] = [];
+        } else {
+          const typedValue = v[1];
+          newSearchAttributes[k.name] = Array.isArray(typedValue) ? typedValue : ([typedValue] as SearchAttributeValue);
+        }
+      }
+      return {
+        ...info,
+        searchAttributes: {
+          ...info.searchAttributes,
+          ...newSearchAttributes,
+        },
+        typedSearchAttributes: info.typedSearchAttributes.updateAttributes(searchAttributes),
+      };
+    });
+  } else {
+    // Legacy search attributes
+    activator.pushCommand({
+      upsertWorkflowSearchAttributes: {
+        searchAttributes: mapToPayloads(searchAttributePayloadConverter, searchAttributes),
       },
-    };
-  });
+    });
+
+    activator.mutateWorkflowInfo((info: WorkflowInfo): WorkflowInfo => {
+      // Use updateSearchAttributes to get a deep copy of the current state.
+      let typedSearchAttributes = info.typedSearchAttributes.updateAttributes([]);
+      for (const [k, v] of Object.entries(searchAttributes)) {
+        const typedKey = TypedSearchAttributes.getKeyFromUntyped(k, v);
+
+        // Unable to discern the typing of the key, skip.
+        if (!typedKey) {
+          continue;
+        }
+
+        // Delete the key if the value is undefined/null/empty list.
+        if (!v || (Array.isArray(v) && v.length === 0)) {
+          const pair = [typedKey, null];
+          if (isTypedSearchAttributeUpdatePair(pair)) {
+            typedSearchAttributes = typedSearchAttributes.updateAttributes([pair]);
+          }
+          continue;
+        }
+
+        let newValue: unknown = v;
+        if (typedKey.type !== SearchAttributeType.KEYWORD_LIST) {
+          // Skip if we found multiple values for a non-list key.
+          if (Array.isArray(v) && v.length > 1) {
+            continue;
+          }
+          // Convert to single-value.
+          newValue = v[0];
+        }
+
+        if (typedKey.type === SearchAttributeType.DATETIME) {
+          newValue = new Date(newValue as string);
+        }
+
+        const typedValue: unknown = [typedKey.type, newValue];
+        const pair: unknown = [typedKey, typedValue];
```

Comment:
That seems incorrect. Both `typedKey` and `typedValue` contains the type.

Reply:
Removed given the different API